testing/test_uncertain_conditional.py

```diff
@@ -0,0 +1,101 @@
+        X_samples = np.random.randn(mc_iter, 1) * (Xstd ** 0.5) + Xmu
+        F_mu, F_var = self.predict_f(X_samples)
+        F_samples = F_mu + np.random.randn(*F_var.shape) * (F_var ** 0.5)
+        mean = np.mean(F_samples, axis=0)
+        covar = np.cov(F_samples.T)
+        return mean, covar
+
+
+class NoUncertaintyTest(GPflowTestCase):
+
+    def setUp(self):
+        self.N, self.Nnew, self.D_out = 7, 2, 3
+        self.X = np.linspace(-5, 5, self.N)[:, None] + np.random.randn(self.N, 1)
+        self.Y = np.hstack([np.sin(self.X), np.cos(self.X), self.X**2])
+        self.Xnew_mu = np.random.randn(self.Nnew, 1)
+        self.Xnew_covar = np.zeros((self.Nnew, 1, 1))
+
+    def test_whiten(self):
+        k = gpflow.ekernels.RBF(1, variance=np.random.randn() ** 2)
+        model = MomentMatchingSVGP(self.X, self.Y, k, gpflow.likelihoods.Gaussian(), Z=self.X.copy(), whiten=True)
+        model.compile()
+        gpflow.train.AdamOptimizer().minimize(model)
+
+        mean1, var1 = model.predict_f(self.Xnew_mu)
+        mean2, var2 = model.uncertain_predict_f_moment_matching(self.Xnew_mu, self.Xnew_covar)
+
+        np.testing.assert_allclose(mean1, mean2)
+        (np.testing.assert_allclose(var1[n,:], np.diag(var2[n,...])) for n in range(self.Nnew))
+
+    def test_non_whiten(self):
+        k = gpflow.ekernels.RBF(1, variance=np.random.randn() ** 2)
+        model = MomentMatchingSVGP(self.X, self.Y, k, gpflow.likelihoods.Gaussian(), Z=self.X.copy(), whiten=False)
+        model.compile()
+        gpflow.train.AdamOptimizer().minimize(model)
+
+        mean1, var1 = model.predict_f(self.Xnew_mu)
+        mean2, var2 = model.uncertain_predict_f_moment_matching(self.Xnew_mu, self.Xnew_covar)
+
+        np.testing.assert_allclose(mean1, mean2)
+        (np.testing.assert_allclose(var1[n,:], np.diag(var2[n,...])) for n in range(self.Nnew))
+
+
+
+class MonteCarloTest(GPflowTestCase):
+
+    def setUp(self):
+        self.N, self.Nnew, self.D_out = 7, 2, 3
+        self.X = np.linspace(-5, 5, self.N)[:, None] + np.random.randn(self.N, 1)
+        self.Y = np.hstack([np.sin(self.X), np.cos(self.X) * 2, self.X**2])
+        self.Xnew_mu = np.random.randn(self.Nnew, 1)
+        self.Xnew_covar = np.random.randn(self.Nnew, 1, 1) ** 2
+
+    def test_whiten(self):
+        k = gpflow.ekernels.RBF(1, variance=np.random.randn() ** 2)
+        model = MomentMatchingSVGP(self.X, self.Y, k, gpflow.likelihoods.Gaussian(), Z=self.X.copy(), whiten=True)
+        model.compile()
+        gpflow.train.AdamOptimizer().minimize(model)
+
+        mean1, var1 = model.uncertain_predict_f_moment_matching(self.Xnew_mu, self.Xnew_covar)
+
+        for n in range(self.Nnew):
+            mean2, var2 = model.uncertain_predict_f_monte_carlo(self.Xnew_mu[n,...], self.Xnew_covar[n,...])
+            np.testing.assert_almost_equal(mean1[n,...], mean2, decimal=3)
+            np.testing.assert_almost_equal(var1[n,...], var2, decimal=2)
```

Comment:
Just a thought: I've been caught out before using MC unit tests where the result is numerically close but still wrong. Really the test should make sure that the convergence is 1/sqrt(S). Numerics can be a pain though when it gets close though